main/main.go

```diff
@@ -0,0 +1,259 @@
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	"github.com/stackrox/k8s-istio-cve-pusher/utils"
+)
+
+const (
+	logDataFeedName      = "data feed name"
+	kubernetesBucketDir  = "k8s"
+	istioBucketDir       = "istio"
+	cveFileExt           = ".json"
+	k8sCveFileName       = "NVDk8sCVEs.json"
+	k8sCveHashFileName   = "k8sCVEsHash"
+	istioCveFileName     = "NVDistioCVEs.json"
+	istioCveHashFileName = "istioCVEsHash"
+)
+
+type component int
+
+const (
+	k8s component = iota
+	istio
+)
+
+type fileAndBucketPaths struct {
+	cveFileNameWithPath     string
+	cveHashFileNameWithPath string
+	cveBucketSubPath        string
+	cveHashBucketSubPath    string
+}
+
+func main() {
+	if err := runCmd(); err != nil {
+		log.Fatal(err)
+		os.Exit(1)
+	}
+}
+
+func runCmd() error {
+	path, err := ioutil.TempDir("", "")
+	if err != nil {
+		return err
+	}
+
+	feedReader, err := getDataFeeds(path)
+	if err != nil {
+		return errors.Wrap(err, "error downloading from NVD DB")
+	}
+
+	k8sCVEs, istioCVEs, err := getK8sAndIstioCVEs(feedReader)
+	if err != nil {
+		return err
+	}
+
+	if err = storeK8sAndIstioCVEsToFile(k8sCVEs, istioCVEs, path); err != nil {
+		return err
+	}
+
+	if err = pushK8sAndIstioCVEsToBucket(path); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func storeK8sAndIstioCVEsToFile(k8sCVEs, istioCVEs []utils.NvdCVEEntry, path string) error {
+	k8sPaths, err := getPaths(path, k8s)
+	if err != nil {
+		return err
+	}
+	if err := utils.StoreCVEsToFile(k8sCVEs, k8sPaths.cveFileNameWithPath, k8sPaths.cveHashFileNameWithPath); err != nil {
+		return err
+	}
+
+	istioPaths, err := getPaths(path, istio)
+	if err != nil {
+		return err
+	}
+	if err := utils.StoreCVEsToFile(istioCVEs, istioPaths.cveFileNameWithPath, istioPaths.cveHashFileNameWithPath); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func pushK8sAndIstioCVEsToBucket(path string) error {
+	if err := pushCVEsToBucket(path, k8s); err != nil {
+		return err
+	}
+
+	if err := pushCVEsToBucket(path, istio); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func pushCVEsToBucket(path string, c component) error {
+	bucketName := os.Getenv(utils.GCloudBucketEnvVar)
+	paths, err := getPaths(path, c)
+	if err != nil {
+		return err
+	}
+
+	if err := utils.WriteToBucket(bucketName, paths.cveBucketSubPath, paths.cveFileNameWithPath); err != nil {
+		return err
+	}
+	log.Infof("done pushing to bucket path %q", filepath.Join(bucketName, paths.cveBucketSubPath))
+
+	if err := utils.WriteToBucket(bucketName, paths.cveHashBucketSubPath, paths.cveHashFileNameWithPath); err != nil {
```

Comment:
hashes need to be pushed first. Reason:
- Case 1: Client reads old feed data and new hash data: client will keep using old feed data indefinitely, until the hashes change again
- Case 2: Client reads old hash and new feed data: client will get new feed data, only downside is that upon next run it downloads data it already has (but this will only happen once for each hash change)